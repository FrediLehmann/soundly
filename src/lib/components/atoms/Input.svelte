<script lang="ts">
  export let name: string;
  export let label: string;
  export let value: any;
  export let error: string = '';

  // Show label only to screen-readers
  // https://tailwindcss.com/docs/screen-readers
  export let labelSrOnly: boolean = false;
</script>

<div class="grid grid-cols-3 w-full">
  <label for={name} class="self-center" class:sr-only={labelSrOnly}>
    {label}
  </label>
  <input
    id={name}
    {...$$restProps}
    bind:value
    class={`${
      $$props.class || ''
    } bg-gray-100 px-3 py-1 focus:outline-none focus:ring focus:border-blue-300 col-span-2`}
    class:col-span-3={labelSrOnly}
  />
  {#if error}
    <span
      class="text-red-500 mt-1 text-xs font-semibold col-start-2 col-span-2"
    >
      {error}
    </span>
  {/if}
</div>
