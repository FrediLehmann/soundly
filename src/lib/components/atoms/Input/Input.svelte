<script lang="ts">
  export let name: string;
  export let label: string;
  export let value: string | number | string[];
  export let error = '';
  export let required = false;

  // Show label only to screen-readers
  // https://tailwindcss.com/docs/screen-readers
  export let labelSrOnly = false;
</script>

<div class="grid grid-cols-3 w-full">
  <label for={name} class="self-center" class:sr-only={labelSrOnly}>
    {label}
    {#if required}
      <span class="ordinal">*</span>
    {/if}
  </label>
  <input
    id={name}
    {...$$restProps}
    bind:value
    class={`${
      $$props.class ? `${$$props.class} ` : ''
    }bg-gray-100 px-3 py-1 focus:outline-none focus:ring focus:border-blue-300 col-span-2`}
    class:col-span-3={labelSrOnly} />
  {#if error}
    <span
      class="text-red-500 mt-1 text-xs font-semibold col-start-2 col-span-2">
      {error}
    </span>
  {/if}
</div>
