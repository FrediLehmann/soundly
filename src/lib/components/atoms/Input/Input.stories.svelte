<script>
  import { Meta, Story, Template } from '@storybook/addon-svelte-csf';
  import { userEvent, within } from '@storybook/testing-library';

  import Input from './Input.svelte';
</script>

<Meta
  title="Atoms/Input"
  component={Input}
  argTypes={{
    name: {
      control: { type: 'text' },
      table: {
        type: {
          summary: 'Input identification'
        }
      }
    },
    label: {
      control: { type: 'text' },
      table: {
        type: {
          summary: 'Input label text'
        }
      }
    },
    value: {
      control: { type: 'text' },
      table: {
        type: {
          summary: 'The value of the input, which can be bound'
        }
      }
    },
    error: {
      control: { type: 'text' },
      table: {
        type: {
          summary: 'Error text, in case there is an validation error'
        }
      }
    },
    required: {
      control: { type: 'boolean' },
      table: {
        type: {
          summary: 'Sets the input as required (showing the * symbol)'
        }
      }
    },
    labelSrOnly: {
      control: { type: 'boolean' },
      table: {
        type: {
          summary:
            'Show or hide the label from the UI without hidding it from screen readers.'
        }
      }
    }
  }} />

<Template let:args>
  <Input {...args} />
</Template>

<Story
  name="With label"
  args={{ name: 'Test', label: 'This Input' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const input = canvas.getByRole('textbox');
    userEvent.type(input, 'Inputting a text, here.');
  }} />

<Story
  name="Disabled"
  args={{ name: 'Test', label: 'This Input', disabled: true }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const input = canvas.getByRole('textbox');
    userEvent.type(input, 'Inputting a text, here.');
  }} />

<Story
  name="With error"
  args={{ name: 'Test', label: 'This Input', error: 'There is an error' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const input = canvas.getByRole('textbox');
    userEvent.type(input, 'Inputting a text, here.');
  }} />

<Story
  name="Required"
  args={{ name: 'Test', label: 'This Input', required: true }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const input = canvas.getByRole('textbox');
    userEvent.type(input, 'Inputting a text, here.');
  }} />

<Story
  name="Hidden label"
  args={{ name: 'Test', label: 'This Input', labelSrOnly: true }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    const input = canvas.getByRole('textbox');
    userEvent.type(input, 'Inputting a text, here.');
  }} />
