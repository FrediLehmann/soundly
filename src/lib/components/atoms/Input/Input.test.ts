/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import { render } from '@testing-library/svelte';

import Input from './Input.svelte';

describe('<Input /> component', () => {
  test('renders correctly', () => {
    let res = render(Input, {
      name: 'test',
      label: 'Test',
      value: 'Test',
      error: 'Test',
      required: true
    });

    expect(res.component).toMatchSnapshot();

    res = render(Input, {
      name: 'test',
      label: 'Test',
      value: 'Test',
      error: 'Test',
      required: true,
      labelSrOnly: true
    });
    expect(res.component).toMatchSnapshot();
  });
});
