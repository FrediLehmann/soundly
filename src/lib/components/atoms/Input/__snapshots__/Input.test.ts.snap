// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> component renders correctly 1`] = `
Input {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      "Test",
      "test",
      "Test",
      "Test",
      true,
      false,
      Object {},
      [Function],
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {
      "error": 3,
      "label": 2,
      "labelSrOnly": 5,
      "name": 1,
      "required": 4,
      "value": 0,
    },
    "root": <div>
      <div
        class="container svelte-1ulq3bj"
        data-required="true"
        data-sr-only="false"
      >
        <label
          class="svelte-1ulq3bj"
          for="test"
        >
          Test
        </label>
         
        <input
          class="svelte-1ulq3bj"
          id="test"
          required=""
        />
         
        <span
          class="error svelte-1ulq3bj"
        >
          Test
        </span>
      </div>
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$$set": [Function],
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;

exports[`<Input /> component renders correctly 2`] = `
Input {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      "Test",
      "test",
      "Test",
      "Test",
      true,
      true,
      Object {},
      [Function],
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {
      "error": 3,
      "label": 2,
      "labelSrOnly": 5,
      "name": 1,
      "required": 4,
      "value": 0,
    },
    "root": <div>
      <div
        class="container svelte-1ulq3bj"
        data-required="true"
        data-sr-only="true"
      >
        <label
          class="svelte-1ulq3bj"
          for="test"
        >
          Test
        </label>
         
        <input
          class="svelte-1ulq3bj"
          id="test"
          required=""
        />
         
        <span
          class="error svelte-1ulq3bj"
        >
          Test
        </span>
      </div>
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$$set": [Function],
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;
